connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java

```diff
@@ -0,0 +1,364 @@
+import java.util.Collections;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.concurrent.ExecutionException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MirrorSourceConnector extends SourceConnector {
+
+    private static final Logger log = LoggerFactory.getLogger(MirrorSourceConnector.class);
+    private static final ResourcePatternFilter ANY_TOPIC = new ResourcePatternFilter(ResourceType.TOPIC,
+        null, PatternType.ANY);
+    private static final AclBindingFilter ANY_TOPIC_ACL = new AclBindingFilter(ANY_TOPIC, AccessControlEntryFilter.ANY);
+
+    private Scheduler scheduler;
+    private MirrorConnectorConfig config;
+    private SourceAndTarget sourceAndTarget;
+    private String connectorName;
+    private TopicFilter topicFilter;
+    private ConfigPropertyFilter configPropertyFilter;
+    private List<TopicPartition> knownTopicPartitions = Collections.emptyList();
+    private Set<String> knownTargetTopics = Collections.emptySet();
+    private ReplicationPolicy replicationPolicy;
+    private int replicationFactor;
+    private AdminClient sourceAdminClient;
+    private AdminClient targetAdminClient;
+
+    public MirrorSourceConnector() {
+        // nop
+    }
+
+    // visible for testing
+    MirrorSourceConnector(SourceAndTarget sourceAndTarget, ReplicationPolicy replicationPolicy,
+            TopicFilter topicFilter) {
+        this.sourceAndTarget = sourceAndTarget;
+        this.replicationPolicy = replicationPolicy;
+        this.topicFilter = topicFilter;
+    } 
+
+    @Override
+    public void start(Map<String, String> props) {
+        config = new MirrorConnectorConfig(props);
+        if (!config.enabled()) {
+            return;
+        }
+        connectorName = config.connectorName();
+        sourceAndTarget = new SourceAndTarget(config.sourceClusterAlias(), config.targetClusterAlias());
+        topicFilter = config.topicFilter();
+        configPropertyFilter = config.configPropertyFilter();
+        replicationPolicy = config.replicationPolicy();
+        replicationFactor = config.replicationFactor();
+        synchronized (this) {
+            sourceAdminClient = AdminClient.create(config.sourceAdminConfig());
+            targetAdminClient = AdminClient.create(config.targetAdminConfig());
+        }
+        scheduler = new Scheduler(MirrorSourceConnector.class);
+        scheduler.execute(this::loadTopicPartitions, "loading initial set of topic-partitions");
+        scheduler.execute(this::createTopicPartitions, "creating downstream topic-partitions");
+        scheduler.scheduleRepeating(this::syncTopicAcls, config.syncTopicAclsInterval(), "syncing topic ACLs");
+        scheduler.scheduleRepeating(this::syncTopicConfigs, config.syncTopicConfigsInterval(),
+            "syncing topic configs");
+        scheduler.scheduleRepeatingDelayed(this::refreshTopicPartitions, config.refreshTopicsInterval(),
```

Comment:
The KIP seems to refresh.topics is true by default. The property REFRESH_TOPICS_ENABLED is named refresh.topics.enabled and the default does not seem to be REFRESH_TOPICS_ENABLED_DEFAULT used.
So the topics are not refreshed by default.

Reply:
The default is set here:

https://github.com/apache/kafka/pull/6295/files#diff-e3868ea5d15e72c75d65799355c2cc97R72

And used here:

https://github.com/apache/kafka/pull/6295/files#diff-e3868ea5d15e72c75d65799355c2cc97R259

I've verified this works as expected. Happy to take suggestions if this is not clear.

Reply:
You're right.

I might miss something obvious, but the propagation of topic creation on downstream cluster does not seem to work.
The MirrorSourceConnector.refreshTopicPartitions detects a new topic properly and and it requests a task reconfiguration.
But the code that really proceed to the topic creation seems to be in MirrorSourceConnector.createTopicPartitions and this task is only executed once at startup and not at the reconfigure phase.

Reply:
@vpernin Connector.reconfigure() by default just calls stop() and start(config). There is no other logic related to reconfiguration at present.

Reply:
I'm just noticing, that in my case, the creation of a new topic (matching replication pattern) upstream is not propagated downstream.
The subject of the reconfigure process is just a attempt to explain the problem.

KMM2 need to be restarted, so MirrorSourceConnector.createTopicPartitions is called.
Connector.reconfigure() does indeed stop and start, which would work, but this is not invoked. We call context.requestTaskReconfiguration() and it does not do that.